attestation-agent/kbs_protocol/src/keypair.rs

```diff
@@ -3,80 +3,160 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-use anyhow::{Context, Result};
+use anyhow::{anyhow, bail, Context, Result};
 
 use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine};
 use crypto::{
+    ec::{Curve, EcKeyPair, KeyWrapAlgorithm},
     rsa::{PaddingMode, RSAKeyPair},
-    WrapType,
 };
-use kbs_types::{Response, TeePubKey};
-use serde::Deserialize;
+use kbs_types::{ProtectedHeader, Response, TeePubKey};
+use log::warn;
 use zeroize::Zeroizing;
 
 #[derive(Clone, Debug)]
 pub struct TeeKeyPair {
-    keypair: RSAKeyPair,
+    key: TeeKey,
+}
+
+#[derive(Clone, Debug)]
+pub enum TeeKey {
+    Rsa(Box<RSAKeyPair>),
+    Ec(Box<EcKeyPair>),
 }
 
 impl TeeKeyPair {
+    /// Create a new Tee key pair. We by default to use EC key pair.
     pub fn new() -> Result<Self> {
-        Ok(Self {
-            keypair: RSAKeyPair::new()?,
-        })
+        let key = TeeKey::Ec(Box::default());
+        Ok(Self { key })
     }
 
     /// Export TEE public key as specific structure.
     pub fn export_pubkey(&self) -> Result<TeePubKey> {
-        let k_mod = URL_SAFE_NO_PAD.encode(self.keypair.n());
-        let k_exp = URL_SAFE_NO_PAD.encode(self.keypair.e());
+        match &self.key {
+            TeeKey::Rsa(key) => {
+                let k_mod = URL_SAFE_NO_PAD.encode(key.n());
+                let k_exp = URL_SAFE_NO_PAD.encode(key.e());
 
-        Ok(TeePubKey::RSA {
-            alg: PaddingMode::PKCS1v15.as_ref().to_string(),
-            k_mod,
-            k_exp,
-        })
+                Ok(TeePubKey::RSA {
+                    alg: PaddingMode::OAEP.as_ref().to_string(),
+                    k_mod,
+                    k_exp,
+                })
+            }
+            TeeKey::Ec(key) => {
+                let x = URL_SAFE_NO_PAD.encode(key.x()?);
+                let y = URL_SAFE_NO_PAD.encode(key.y()?);
+
+                Ok(TeePubKey::EC {
+                    crv: Curve::P256.as_ref().to_string(),
+                    alg: KeyWrapAlgorithm::EcdhEsA256Kw.as_ref().to_string(),
+                    x,
+                    y,
+                })
+            }
+        }
     }
 
     #[inline]
-    pub fn decrypt(&self, mode: PaddingMode, cipher_text: Vec<u8>) -> Result<Vec<u8>> {
-        self.keypair.decrypt(mode, cipher_text)
+    pub fn unwrap_cek(&self, header: &ProtectedHeader, cipher_text: Vec<u8>) -> Result<Vec<u8>> {
```

Comment:
If this is now called `unwrap_cek` maybe we should change the `cipher_text` arg to be `wrapped_cek`.

Reply:
Nice idea.